### PR TITLE
[pythonic resources] Attach resource function name to function-style resources in snapshots

### PR DESCRIPTION
**Pull request by benpankow**
## Summary

Extracts the wrapped resource function name when getting a resource name for snapshot purposes. Previously, the type would just display as `ResourceDefinition` or `IOManagerDefinition`.

Screen Shot 2023-05-05 at 11 26 10 AM

## Test Plan

New unit tests, tested locally.

### CODE REVIEW

**Comment by benpankow** [comment]
Current dependencies on/for this PR:
* master
  * **PR #14132** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/14132" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/14132?utm_source=stack-comment).

**Review by jamiedemaria — APPROVED** [review]
seems good to me, but would be good to get approval from nick or yuhan too

**Review by yuhan — APPROVED** [review]
this looks good me. lets land this to unblock the dependent tracking work while ben's out

additionally, would be nice to add a couple unit tests in integration libs to make sure this works as expected - this can be in a separate pr.

**Pull request merged**
